File: 백준/Silver/15654. N과 M （5）/N과 M （5）.py
import sys
input = sys.stdin.readline

N, M = map(int, input().split())

arr = list(map(int, input().split()))
arr.sort()

ans = []
def back(start):
  if len(ans) == M:
    print(' '.join(map(str, ans)))
    return
  
  for i in range(len(arr)):
    if arr[i] in ans:
      continue
    ans.append(arr[i])
    back(start + 1)
    ans.pop()

back(0)

File: 백준/Silver/15654. N과 M （5）/N과 M （5）.java
import java.util.*;
import java.util.stream.Collector;

public class Main {

  static int N, M;
  static List<Integer> arr = new ArrayList<>();
  static List<Integer> ans = new ArrayList<>();
  
  public static void main(String[] args) {

    Scanner in = new Scanner(System.in);
    N = in.nextInt();
    M = in.nextInt();

    for (int i = 0; i < N; i++) {
      arr.add(in.nextInt());
    }
    Collections.sort(arr);

    back(0);

  }

  static void back(int start) {
    if (ans.size() == M) {
      for (int a: ans) {
        System.out.print(a + " ");
      }
      System.out.println();
      return;
    }

    for (int i = 0; i < N; i++) {
      if (ans.contains(arr.get(i))) {
        continue;
      }

      ans.add(arr.get(i));
      back(start + 1);
      ans.remove(ans.size() - 1);
    }
  }

}